Add component feature for plotting vectors

MatthewFlamm opened the pull request:
Fixes title of scalar bar in `BasePlotter`. It looks like `title` was meant to be used, but `stitle` was used instead.

Confusingly, it looks like add_volume is intended to use `stitle`?

Example:
```py
import pyvista as pv
from pyvista import examples
mesh = examples.download_carotid().threshold(145, scalars="scalars")
p = pv.Plotter()
p.add_mesh(mesh, scalars="vectors")
p.show()
```

Old behavior:
![image](https://user-images.githubusercontent.com/39341281/72540456-d4fa4f00-384e-11ea-8519-5a3a41b26466.png)


New Behavior:
![image](https://user-images.githubusercontent.com/39341281/72540388-bc8a3480-384e-11ea-80df-88a168cbba23.png)

MatthewFlamm commented:
This PR also partially (or fully?) addresses the plotting vector documentation issue as the scalar bar title now displays '-normed' which makes the default behavior more self explanatory.

MatthewFlamm commented:
Not sure why this test is failing consistently: test_default_name_tracking.

GuillaumeFavelier commented:
The scalar bar is added to the list of actors now that's why you notice `assert 101 == 100` in [Travis](https://travis-ci.org/pyvista/pyvista/jobs/638018887#L1157).

You can try modifying the test with: `add_mesh(show_scalar_bar=False)`

banesullivan commented:
I don't think we should modify that test. A scalar bar shouldn't be added in that routine because no scalars are shown in that example as the `color` argument is given:

https://travis-ci.org/pyvista/pyvista/jobs/638018887#L1153

MatthewFlamm commented:
I agree @banesullivan .  I need to look deeper at how the title is created with color.  Maybe additional tests need to be added with other scalars bar modifying options too.  This was a helpful test.

MatthewFlamm commented:
It looks like the scalars bar title has multiple problems.  It looks like the only way the behavior with a color created no scalar bar, was an unintended consequence of the error I'm trying to fix.  Right now I suspect that the last if statement should have rgb_color instead of rgb but the workings of this function are very interconnected.  A small change in one spot makes a huge change somewhere else.

MatthewFlamm commented:
I changed to rgb_color and the tests are passing. But this part of the if statement `(not rgb_color or _custom_opac)` I'm not sure about. I would assume if the color is specified you don't want a scale bar even if the opacity is changing.  Worth another consideration given the other issues.

banesullivan commented:
> Right now I suspect that the last if statement should have rgb_color instead of rgb but the workings of this function are very interconnected. A small change in one spot makes a huge change somewhere else.

Yeah, the `add_mesh` and `add_volume` methods have a really high complexity which we've been meaning to simplify... Alex and I are really the only ones that have written those methods and both of us are stressed for time to do a refactor like outlined in https://github.com/pyvista/pyvista/issues/299. This might require me to sit down with some coffee and reimplement the title implementation for those methods.


> I changed to rgb_color and the tests are passing. But this part of the if statement (not rgb_color or _custom_opac) I'm not sure about. I would assume if the color is specified you don't want a scale bar even if the opacity is changing. Worth another consideration given the other issues.

Hm... I will need to look this over carefully and make sure everything is behaving as intended (and make sure `add_volume` is doing the same thing. This might take a while before I have the time.

MatthewFlamm commented:
I made this change so I could change the title for plotting components of vectors.  Would it be okay to forge ahead with adding components and add the title in the current broken form, or should I hold off in general?

banesullivan commented:
If you want to give it a try, go ahead on this PR! Either way, I will have to do a critical review because of how crazy/complicated the `add_mesh` method has become. If any issue arises from this implementation, I can resolve them and use your work

MatthewFlamm commented:
`rgb_color` is also set in multiple points in the code, which broke the usage of the normed in the vector plotting scalar bar.  I switched yet again to use `color` there.

I also added the functionality for the component feature here as requested.  Using `component` parameter will allow for choosing to plot only a single component of a vector in `add_mesh`.  It also modifies the title to indicate which component is plotted.

The new functionality at least shouldn't add to the interconnectedness of this function.

Example:
```py
import pyvista as pv
from pyvista import examples
import numpy as np
data = examples.load_uniform()
vector_values = np.empty((data.n_points, 3))
vector_values[:, :] = [3., 4., 0.]
data['vector_values'] = vector_values

p=pv.Plotter()
p.add_mesh(data, scalars='vector_values', clim=[0, 5])
p.show()

p=pv.Plotter()
p.add_mesh(data, scalars='vector_values', component=0, clim=[0, 5])
p.show()

p=pv.Plotter()
p.add_mesh(data, scalars='vector_values', component=2, clim=[0, 5])
p.show()
```

![image](https://user-images.githubusercontent.com/39341281/72618504-37178a80-3909-11ea-9086-efb76eb656b5.png)
![image](https://user-images.githubusercontent.com/39341281/72618520-40085c00-3909-11ea-9d1c-793a6642ac62.png)
![image](https://user-images.githubusercontent.com/39341281/72618532-4991c400-3909-11ea-9546-6849413a1902.png)

MatthewFlamm commented:
You could also apply the vector component feature separately (let me know and I can submit a separate PR), but the scalar bar title won't work correctly.  But, I'm fine with the plan to leave this as a reference pending a better structure for add_mesh.

banesullivan commented on pyvista/plotting/plotting.py:
```diff
@@ -1293,8 +1310,8 @@ def prepare_mapper(scalars):
-        if stitle is not None and show_scalar_bar and (not rgb or _custom_opac):
-            self.add_scalar_bar(stitle, **scalar_bar_args)
+        if title is not None and show_scalar_bar and (not color or _custom_opac):
+            self.add_scalar_bar(title, **scalar_bar_args)
```
This change is incorrect and not needed for the desired results. This change adds a scalar bar when one is not supped to be there. Also, the docs are failing for me locally (I'm not sure why Travis didn't pick that up 😨 )

```py
from pyvista import examples
examples.load_airplane().plot()
```

![download](https://user-images.githubusercontent.com/22067021/73333259-78bc0580-4226-11ea-973e-9b32f3a6e12a.png)

MatthewFlamm replied:
hmmm it seems that any of the versions of my edits and the original behavior all have some issue.  Should all the edits here be reverted?  The naming of the vector norm and components will not work, but that is the current state.

MatthewFlamm replied:
I'm now thinking that `title` was correct here and the `stitle` in the vector valued portion should be changed to `title`.  But, every other thought I've had was wrong.  I'm going to suggest that all of the scalar bar title commits be reverted back to the original state.  Then I'll open a new issue to track this issue.

banesullivan commented on pyvista/plotting/plotting.py:
```diff
@@ -1133,8 +1137,21 @@ def add_mesh(self, mesh, color=None, style=None, scalars=None,
+                             'dimensionality of the scalars array: {}').format(
+                                 scalars.shape[1]
+                             )
+                        )
```
Nice!

banesullivan commented:
Also, the `vtkDataSetMapper` class has the [`SetArrayComponent`](https://vtk.org/doc/nightly/html/classvtkMapper.html#acb151c4a79b4cf87205749d52e46b02b) method which I tried using to avoid copying the component of the array to a new array, but I couldn't get it to work... not sure what the method does.

MatthewFlamm commented:
Based on the comments in https://github.com/pyvista/pyvista/pull/570#issuecomment-579790574, I'd also like to add a test to make sure that the dataset object is not modified by taking the norm or plotting the individual components.  There are probably several cases to consider here with different ways of passing in the vector-valued scalar.

MatthewFlamm commented:
I reverted all the changes to the title of the scalar bar, and added to test to make sure the operations for vector valued scalars do not change any of the inputs.

With the current state of this PR, this is what it looks like using the same example above.  Note that the titles of the scalar bars do not work correctly, but I'm hesitant to keep tinkering with it because every change has unintended side effects.

In the docstring of `add_mesh`, I still don't like 'vector valued scalars', but I'm not sure what else to call it.

![image](https://user-images.githubusercontent.com/39341281/73754461-711ec400-4732-11ea-9056-e6a1ae18b553.png)

![image](https://user-images.githubusercontent.com/39341281/73754513-7f6ce000-4732-11ea-9c23-a8a2249405b3.png)

![image](https://user-images.githubusercontent.com/39341281/73754555-8dbafc00-4732-11ea-9932-b6cd6b13006a.png)
